src/codegen/compile.ml

```diff
@@ -1574,58 +1668,223 @@ module BitTagged = struct
+  let untag line env pty = Type.(match pty with
+    | Nat | Int | Int64 | Int32 ->
+      let ubitsl = Int32.of_int (ubits_of pty) in
+      sanity_check_tag line env pty ^^
+      compile_shrS_const (Int32.sub 32l ubitsl) ^^
+      G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32))
+    | Nat64 | Nat32 ->
+      let ubitsl = Int32.of_int (ubits_of pty) in
+      sanity_check_tag line env pty ^^
+      compile_shrU_const (Int32.sub 32l ubitsl) ^^
+      G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32))
+    | _ -> assert false)
 
   (* 32 bit numbers, dynamic, w.r.t `Int` *)
 
-  let if_can_tag_i32 env retty is1 is2 =
-    Func.share_code1 Func.Never env "cannot_tag_i32" ("x", I32Type) [I32Type] (fun env get_x ->
-      (* checks that all but the low 30 bits are both either 0 or 1 *)
-      get_x ^^ compile_shrU_const 30l ^^
-      G.i (Unary (Wasm.Values.I32 I32Op.Popcnt)) ^^
-      G.i (Unary (Wasm.Values.I32 I32Op.Ctz))
-    ) ^^
-    E.if_ env retty is1 is2
-
-  let if_can_tag_u32 env retty is1 is2 =
-    compile_shrU_const 30l ^^
-    E.if_ env retty is2 is1 (* NB: swapped branches *)
+  let sanity_check_can_tag_i32 env pty get_x =
+    if TaggingScheme.debug || !Flags.sanity then
+      get_x ^^
+      Func.share_code2 Func.Always env (prim_fun_name pty "check_can_tag_i32") (("res", I32Type), ("x", I32Type)) [I32Type]
+        (fun env get_res get_x ->
+          let lower_bound, upper_bound = Type.(
+            match pty with
+            |  Nat | Int | Int64 | Int32 ->
+              let sbits = sbits_of pty in
+              (Int32.(neg (shift_left 1l sbits)), Int32.shift_left 1l sbits)
+            |  Nat64 | Nat32 ->
+              let ubits = ubits_of pty in
+              (0l, Int32.shift_left 1l ubits)
+            | _ -> assert false)
+          in
+          (* lower_bound <= x < upper_bound *)
+          compile_unboxed_const lower_bound ^^
+          get_x ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.LeS)) ^^
+          get_x ^^ compile_unboxed_const upper_bound ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.And)) ^^
+          get_res ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
+          E.else_trap_with env (prim_fun_name pty "check_can_tag_i32") ^^
+          get_res)
+    else
+      G.nop
 
-  let tag_i32 = compile_shl_const 1l
-  let untag_i32 = compile_shrS_const 1l
+  let if_can_tag_i32 env pty retty is1 is2 = Type.(match pty with
+    |  Nat | Int | Int64 | Int32 ->
+      Func.share_code1 Func.Never env
+        (prim_fun_name pty "if_can_tag_i32") ("x", I32Type) [I32Type] (fun env get_x ->
+          (* checks that all but the low sbits are both either 0 or 1 *)
+          get_x ^^
+          get_x ^^ compile_shrS_const (Int32.of_int ((32 - ubits_of pty) - 1)) ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.Xor)) ^^
+          compile_shrU_const (Int32.of_int (sbits_of pty)) ^^
+          G.i (Test (Wasm.Values.I32 I32Op.Eqz)) ^^
```

Comment:
Just a detail: As most of the logic is symmetric to the 64-bit tagging except for this case, I wonder whether we could apply this logic also to the 64-bit signed tagging test.

Reply:
Done